presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java

```diff
@@ -0,0 +1,101 @@
+import org.apache.hudi.common.table.view.FileSystemViewManager;
+import org.apache.hudi.common.table.view.HoodieTableFileSystemView;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Optional;
+
+import static com.facebook.presto.hive.HiveFileInfo.createHiveFileInfo;
+import static com.facebook.presto.hive.HiveSessionProperties.isHudiMetadataVerificationEnabled;
+import static com.facebook.presto.hive.HiveSessionProperties.isPreferMetadataToListHudiFiles;
+
+public final class HudiDirectoryLister
+        implements DirectoryLister
+{
+    private HoodieTableFileSystemView fileSystemView;
+
+    public HudiDirectoryLister(Configuration conf,
+                               ConnectorSession session,
+                               Table table)
+    {
+        HoodieTableMetaClient metaClient = new HoodieTableMetaClient(conf, table.getStorage().getLocation());
+        this.fileSystemView = FileSystemViewManager.createInMemoryFileSystemView(metaClient,
+                isPreferMetadataToListHudiFiles(session), isHudiMetadataVerificationEnabled(session));
+    }
+
+    @Override
+    public Iterator<HiveFileInfo> list(ExtendedFileSystem fileSystem,
+                                       Table table,
+                                       Path path,
+                                       NamenodeStats namenodeStats,
+                                       PathFilter pathFilter,
+                                       HiveDirectoryContext hiveDirectoryContext)
+    {
+        return new HiveFileIterator(
+                path,
+                p -> new HudiFileInfoIterator(fileSystem, fileSystemView, table.getStorage().getLocation(), p),
+                namenodeStats,
+                hiveDirectoryContext.getNestedDirectoryPolicy(),
+                pathFilter);
+    }
+
+    public static class HudiFileInfoIterator
+            implements RemoteIterator<HiveFileInfo>
+    {
+        private final ExtendedFileSystem fileSystem;
+        private final Iterator<HoodieBaseFile> hoodieBaseFileIterator;
+
+        public HudiFileInfoIterator(ExtendedFileSystem fileSystem,
+                                    HoodieTableFileSystemView fileSystemView,
+                                    String tablePath,
+                                    Path directory)
+        {
+            this.fileSystem = fileSystem;
+            String partition = FSUtils.getRelativePartitionPath(new Path(tablePath), directory);
+            this.hoodieBaseFileIterator = fileSystemView.getLatestBaseFiles(partition).iterator();
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            return hoodieBaseFileIterator.hasNext();
+        }
+
+        @Override
+        public HiveFileInfo next() throws IOException
+        {
+            FileStatus fileStatus = hoodieBaseFileIterator.next().getFileStatus();
+            LocatedFileStatus hoodieFileStatus = new LocatedFileStatus(fileStatus,
```

Comment:
so this will do one additional RPC per base file? if so, would n't this be bad actually for listing performance?

Reply:
@umehrot2 what I meant was if we can simply do something like 

```
LocatedFileStatus hoodieFileStatus = new LocatedFileStatus(fileStatus, 
   new BlockLocation[] {new BlockLocation(name, host, 0, file.getLen())});
```
if so, would this work for hdfs.

Reply:
@umehrot2 do you have an easy test environment for HDFS?

Reply:
Uber and Facebook both use hdfs a lot. So this will come up in the review upstream for sure.

Reply:
Yeah I will make that change, and do some testing on EMR cluster. EMR clusters all have HDFS so I should be able to test.